---
title: 'VISA report '
author: "Guillaume Corre"
date: '`r format(Sys.time())`'
output:
  html_document:
    fig_height: 8
    fig_width: 8
    number_sections: yes
    theme: journal
    toc: yes
---
#Sample name: `r Sample_name`  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,message = FALSE,warning = FALSE)
knitr::opts_knit$set(root.dir = args[1])
```

# Summary Table
```{r}
kable(sumtb,align = "c")
```


# Demultiplexing metrics
Raw reads count and percentage for each detected TAG.  
```{r}
gg1 <- ggplot(demux %>% 
                filter(TAG!="total", Count>0)%>% 
                mutate(TAG = factor(TAG,levels=TAG)),
              aes(TAG,y=Count)) + 
  geom_bar(stat="identity") + 
  theme_bw() + 
  geom_text(aes(label = paste(formatC(Count/sum(Count)*100,
                                      digits=2,
                                      big.mark = ",",format="g"),"%")),
                position=position_dodge(width=1), vjust=-0.25) + 
  scale_y_log10()

gg1
rm(demux)
```

# Provirus, LTR primer and Elongation, GCA starting and Linker trimming

```{r}


q1 <- ggplot(data = all_counts %>% 
               filter(str_detect(Step,"^0[0-9]"),str_detect(description,pattern = "^R1")), 
             aes(Step,reads,fill = str_extract(description,pattern = "Negative$|Positive$|mapped$|unmapped$|2short$|filtered$"))) + 
  geom_bar(stat="identity") + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45,hjust=1,vjust=1)) + 
  scale_fill_discrete("") 
q1
```


# Pipeline metrics

```{r}
kable(all_counts %>% 
        select(Step,description,parent,reads,pct_parent,pct_total) %>% 
        arrange(Step) %>% 
        filter(str_detect(description,pattern = "^R1|^Qua"))%>%
        mutate(reads = formatC(reads,big.mark = ",",format = "d")))
```

![](`r paste(args[1],"pipeline_annotated.png",sep="/")`)


# Annotation
```{r}
full_annot$Chr_IS <- factor(full_annot$Chr_IS,levels = c(paste("chr",1:22,sep = ""),"chrX","chrY","chrM"))
gg1 <- ggplot(full_annot,aes(Chr_IS, fill = Strand_IS)) + 
  geom_bar(aes(y = (..count..)/sum(..count..)),col = "black")  + 
  theme_bw()+
  scale_y_continuous(labels=percent,name = "Percent of IS") +xlab("") +
  coord_polar() + 
  theme(legend.position="top")

```


```{r,fig.height=6,fig.width=12}

gg2 <- ggplot(full_annot, aes(Position))+ 
  geom_bar(aes(y = (..count..)/sum(..count..)))  + 
  theme_bw() + 
  scale_y_continuous(labels=percent,name = "Percent of IS",limits=c(0,1)) + 
  xlab("")

grid.arrange(gg1,gg2,ncol=2)

```


# Wordcloud of most targeted genes
```{r}

aa <- full_annot %>% filter(Regions=="Intragenic") %>% group_by(GeneSymbol) %>% summarise(count = n())

wordcloud(aa$GeneSymbol,aa$count,min.freq = 50,max.words = 200,random.order = F,random.color = F,colors = matlab.like2(50), rot.per = 0.2,scale = c(3,.1))
```



# Relative abundance distribution based on read count
```{r}
UniqueIS <- full_annot %>% select(1:7) %>% distinct() %>% arrange(desc(ReadCount_IS))
require(vegan)

plot(radfit(UniqueIS$ReadCount_IS))

mosaicplot(xtabs(data = All_annot,~(Strand_IS)+ (Strand_Gene),drop.unused.levels = T), col = 1:3, shade = T, main ="", cex = 2,off = 10)

tt <- UniqueIS %>% cast(~Chr_IS + Start_IS + Strand_IS) %>% select(-1)
rarecurve(tt,step=1000,lty = 2)

```

# SonicLength correction

```{r}
plot(as.integer(full_annot$Fragments_Corrected_quant),as.integer(full_annot$SonicAbundance_quanti))
abline(a = 0,b=1)
```


# Relative abundance distribution based on fragment count (sonicLength)
```{r}
UniqueIS <- full_annot %>%
  select(1:6,20) %>% 
  distinct() %>% 
  filter(!is.na(SonicAbundance_quanti)) %>% 
  arrange(desc(SonicAbundance_quanti))

plot(radfit(as.integer(UniqueIS$SonicAbundance_quanti)))

```



```{r}

top <- full_annot %>% 
  group_by(Name_IS) %>% 
  arrange(desc(SonicAbundance_quanti)) %>% 
  ungroup()  %>% 
  slice(1:10)



ggplot(Corrected_FragLengthCount %>%  
  semi_join(top, by = c("IS" = "Name_IS"))%>%
  mutate(IS = factor(IS,levels = top$Name_IS)), 
       aes(Size,ReadCount)) + 
  geom_point() + 
  theme_bw() + 
  facet_wrap(~ IS,scales = "free_y") + 
  scale_y_log10(limits=c(1,NA)) +
  xlab("Fragment length (bp)") + ggtitle("Reads count ~ fragment length for top10 most abundant IS")

```


```{r}

ggplot(Corrected_FragLengthCount %>% 
         group_by(IS) %>% 
         summarize(Frag=n(),ReadCOunt=sum(ReadCount)), 
       aes(ReadCOunt,Frag)) + 
  geom_point() + 
  theme_bw() + 
  xlab ("Read count per IS") +
  ylab("Fragment count per IS") +
  scale_x_log10() + 
  scale_y_log10()

```

# Diversity indices
```{r}
input_mat <-as.matrix(full_annot %>% filter(!is.na(SonicAbundance_quanti))%>% pull(SonicAbundance_quanti))
colnames(input_mat) <- Sample_name


plot(Lc(full_annot %>% filter(!is.na(SonicAbundance_quanti))%>% pull(SonicAbundance_quanti)),col="red",lwd=2)

results2 <- data.frame(diverse::diversity(data = t(input_mat),base = exp(1),type = c("variety","entropy","ev","gs","s")),
                      "Gini_ineq"=ineq(input_mat),
                              "Shannon_ChaoShen2003"= entropart::Shannon(t(input_mat),Correction = "ChaoShen"),
                              "Shannon_MillerMadow1955"= entropart::Shannon(t(input_mat),Correction = "Miller"),
                              "Shannon_GenCov"= entropart::Shannon(t(input_mat),Correction = "GenCov"),
                              "Shannon_Grassberger2003"= entropart::Shannon(t(input_mat),Correction = "Grassberger2003"),
                              "Shannon_ChaoWangJost2013"= entropart::Shannon(t(input_mat),Correction = "ChaoWangJost"),
                              "Shannon_Zhang2012"= entropart::Shannon(t(input_mat),Correction = "ZhangHz"),
                              "Shannon_MillerMadow1955"=entropy::entropy(t(input_mat),method = c("MM"),unit="log"),
                              "Shannon_ChaoChen2003" = entropy::entropy(t(input_mat),method = c("CS"),unit="log"),
                              "Cov_GoodTuring1953" = entropart::Coverage(Ns = t(input_mat),Estimator = "Turing"),
                              "Cov_ChaoShen2010" = entropart::Coverage(Ns = t(input_mat),Estimator = "Chao"),
                              "Cov_ZhangHuang2007" = entropart::Coverage(Ns = t(input_mat),Estimator = "ZhangHuang"),
                      "Rich_Chao1_1984" = entropart::Richness(t(input_mat),Correction = "Chao1"),
                      "Rich_iChao1_2014" = entropart::Richness(t(input_mat),Correction = "iChao1"),
                      #"Rich_Jack" = entropart::Richness(t(input_mat),Correction = "Jackknife"),
                      data.frame(t(estimateR(round(t(input_mat))))),
                      "UC50berry2016" = length(which(cumsum(sort(t(input_mat) / sum(t(input_mat)),decreasing = F))>0.5)),
                      "UC50berry2016norm" = length(which(cumsum(sort(t(input_mat) / sum(t(input_mat)),decreasing = F))>0.5))/length(t(input_mat))*100
                      )
kable(t(results2),digits=3,format.args = list(big.mark=" ",drop0trailing=T,scientific=F))
```


# Session information

```{r}
sessionInfo()
```

---

# Bibliography
```

# function "shannon(Correction=ChaoShen)" from entropart package uses the Coverage estimator from ZhangHuang not ChaoShen

# Cov_ChaoShen2010
# Chao, A. and Shen, T.-J. (2010). Program SPADE: Species Prediction And Diversity Estimation. Program and user's guide. CARE, Hsin-Chu, Taiwan.

# Cov_GoodTuring1953
# Good, I. J. (1953). On the Population Frequency of Species and the Estimation of Population Parameters. Biometrika 40(3/4): 237-264.

# Cov_ZhangHuang2007
# Zhang, Z. and Huang, H. (2007). Turing's formula revisited. Journal of Quantitative Linguistics 14(2-3): 222-241.

# Rich_Chao1_1984
# Chao, A. (1984) Nonparametric estimation of the number of classes in a population. Scandinavian Journal of Statistics 11:265-270.

# Rich_iChao1_2014
# Chiu, C.-H., Wang, Y.-T., Walther, B. A., Chao, A. (2014) An Improved Nonparametric Lower Bound of Species Richness via a Modified Good-Turing Frequency Formula. Biometrics 70(3):671-682

# Shannon_ChaoShen2003
# Chao, A. and Shen, T. J. (2003). Nonparametric estimation of Shannon's index of diversity when there are unseen species in sample. Environmental and Ecological Statistics 10(4): 429-443

# Shannon_ChaoWangJost2013 
# Chao, A., Wang, Y. T. and Jost, L. (2013). Entropy and the species accumulation curve: a novel entropy estimator via discovery rates of new species. Methods in Ecology and Evolution 4(11):1091-1100.

# Shannon_MillerMadow1955
# Miller, G. (1955) Note on the bias of information estimates. In: Quastler, H., editor. Information Theory in Psychology: Problems and Methods: 95-100.

# entropy
# Shannon, C. E. (1948). A Mathematical Theory of Communication. The Bell System Technical Journal 27: 379-423, 623-656.

# Shannon_Grassberger2003
# Grassberger, P. (2003). Entropy Estimates from Insufficient Samplings. ArXiv Physics e-prints 0307138.

# Shannon_GenCov
# Chao, A., Hsieh, T. C., Chazdon, R. L., Colwell, R. K., Gotelli, N. J. (2015) Unveiling the Species-Rank Abundance Distribution by Generalizing Good-Turing Sample Coverage Theory. Ecology 96(5): 1189-1201.

# Shannon_Zhang2012
# Zhang, Z. (2012). Entropy Estimation in Turing's Perspective. Neural Computation 24(5): 1368-1389

#Horvitz, D.G., and D. J. Thompson. 1952. A generalization of sampling without replacement from a finite universe. J. Am. Stat. Assoc. 47:663-685.
```
