---
title: 'VISA report '
author: "Guillaume Corre"
date: '`r format(Sys.time())`'
output:
  html_document:
    fig_height: 8
    fig_width: 8
    number_sections: yes
    theme: journal
    toc: yes
---
#Sample name: `r Sample_name`  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,message = FALSE,warning = FALSE)
knitr::opts_knit$set(root.dir = args[1])
```

# Summary Table
```{r}
kable(sumtb,align = "c")
```


# Demultiplexing metrics
Raw reads count and percentage for each detected TAG.  
```{r}
gg1 <- ggplot(demux %>% 
                filter(TAG!="total", Count>0)%>% 
                mutate(TAG = factor(TAG,levels=TAG)),
              aes(TAG,y=Count)) + 
  geom_bar(stat="identity") + 
  theme_bw() + 
  geom_text(aes(label = paste(formatC(Count/sum(Count)*100,
                                      digits=2,
                                      big.mark = ",",format="g"),"%")),
                position=position_dodge(width=1), vjust=-0.25) + 
  scale_y_log10()

gg1
rm(demux)
```

# Provirus, LTR primer and Elongation, GCA starting and Linker trimming

```{r}


q1 <- ggplot(data = all_counts %>% 
               filter(str_detect(Step,"^0[0-9]"),str_detect(description,pattern = "^R1")), 
             aes(Step,reads,fill = str_extract(description,pattern = "Negative$|Positive$|mapped$|unmapped$|2short$|filtered$"))) + 
  geom_bar(stat="identity") + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45,hjust=1,vjust=1)) + 
  scale_fill_discrete("") 
q1
```


# Pipeline metrics

```{r}
kable(all_counts %>% 
        select(Step,description,parent,reads,pct_parent,pct_total) %>% 
        arrange(Step) %>% 
        filter(str_detect(description,pattern = "^R1|^Qua"))%>%
        mutate(reads = formatC(reads,big.mark = ",",format = "d")))
```

![](`r paste(args[1],"pipeline_annotated.png",sep="/")`)


# Annotation
```{r}
full_annot$Chr_IS <- factor(full_annot$Chr_IS,levels = c(paste("chr",1:22,sep = ""),"chrX","chrY","chrM"))
gg1 <- ggplot(full_annot,aes(Chr_IS, fill = Strand_IS)) + 
  geom_bar(aes(y = (..count..)/sum(..count..)),col = "black")  + 
  theme_bw()+
  scale_y_continuous(labels=percent,name = "Percent of IS") +xlab("") +
  coord_polar() + 
  theme(legend.position="top")

```


```{r,fig.height=6,fig.width=12}

gg2 <- ggplot(full_annot, aes(Position))+ 
  geom_bar(aes(y = (..count..)/sum(..count..)))  + 
  theme_bw() + 
  scale_y_continuous(labels=percent,name = "Percent of IS",limits=c(0,1)) + 
  xlab("")

grid.arrange(gg1,gg2,ncol=2)

```


# Wordcloud of most targeted genes
```{r}

aa <- full_annot %>% filter(Regions=="Intragenic") %>% group_by(GeneSymbol) %>% summarise(count = n())

wordcloud(aa$GeneSymbol,aa$count,min.freq = 50,max.words = 200,random.order = F,random.color = F,colors = matlab.like2(50), rot.per = 0.2,scale = c(3,.1))
```



# Relative abundance distribution based on read count
```{r}
UniqueIS <- full_annot %>% select(1:7) %>% distinct() %>% arrange(desc(ReadCount_IS))
require(vegan)

plot(radfit(UniqueIS$ReadCount_IS))

mosaicplot(xtabs(data = All_annot,~(Strand_IS)+ (Strand_Gene),drop.unused.levels = T), col = 1:3, shade = T, main ="", cex = 2,off = 10)

tt <- UniqueIS %>% cast(~Chr_IS + Start_IS + Strand_IS) %>% select(-1)
rarecurve(tt,step=1000,lty = 2)

```



# Relative abundance distribution based on fragment count
```{r}
UniqueIS <- full_annot %>%
  select(1:6,17) %>% 
  distinct() %>% 
  filter(!is.na(Fragments_Corrected)) %>% 
  arrange(desc(Fragments_Corrected))

plot(radfit(UniqueIS$Fragments_Corrected))

```



```{r}

top <- full_annot %>% 
  group_by(Name_IS) %>% 
  arrange(desc(Fragments_Corrected)) %>% 
  ungroup()  %>% 
  slice(1:10)



ggplot(Corrected_FragLengthCount %>%  
  semi_join(top, by = c("IS" = "Name_IS"))%>%
  mutate(IS = factor(IS,levels = top$Name_IS)), 
       aes(Size,ReadCount)) + 
  geom_point() + 
  theme_bw() + 
  facet_wrap(~ IS,scales = "free_y") + 
  scale_y_log10(limits=c(1,NA)) +
  xlab("Fragment length (bp)") + ggtitle("Reads count ~ fragment length for top10 most abundant IS")

```


```{r}


ggplot(Corrected_FragLengthCount %>% group_by(IS) %>% summarize(Frag=n(),ReadCOunt=sum(ReadCount)), aes(ReadCOunt,Frag)) + geom_point() + theme_bw() + xlab ("Read count per IS") + ylab("Fragment count per IS") + scale_x_log10() + scale_y_log10()

```

# Session information

```{r}
sessionInfo()
```



